py/surveyqa/core.py

```diff
@@ -50,6 +52,57 @@ def check_offline_files(dir):
 
         print("Downloaded offline Bokeh files")
 
+def generate_js(outdir, nights, subset):
+    '''
+    Generates linking.js, which helps in linking all the nightly htmls together
+
+    Args:
+        outdir : directory to write linking.js and to check for previous html files
+        nights : list of nights (strings) to link together
+        subset : if True : nights is a subset, and we need to include all existing html files in outdir
+                 if False : nights is not a subset, and we do not need to include existing html files in outdir
+
+    Writes outdir/linking.js
+    '''
+    f = []
+    f += nights
+    if subset:
+        f_existing = []
+        for (dirpath, dirnames, filenames) in walk(outdir):
+            f.extend(filenames)
+            break
+        regex = re.compile("night-[0-9]+.html")
+        f_existing = [filename for filename in f if regex.match(filename)]
+        f_existing = [i[6:14] for i in f_existing]
+        f += f_existing
+        f = list(dict.fromkeys(f))
+        f.sort()
+    file_js = dict()
+    file_js["first"] = "night-"+f[0]+".html"
+    file_js["last"] = "night-"+f[len(f)-1]+".html"
+
+    for i in np.arange(len(f)):
+        inner_dict = dict()
+        if (len(f) == 1):
+            inner_dict["prev"] = "night-"+f[i]+".html"
+            inner_dict["next"] = "night-"+f[i]+".html"
+        elif i == 0:
+            inner_dict["prev"] = "night-"+f[i]+".html"
+            inner_dict["next"] = "night-"+f[i+1]+".html"
+        elif i == len(f)-1:
+            inner_dict["prev"] = "night-"+f[i-1]+".html"
+            inner_dict["next"] = "night-"+f[i]+".html"
+        else:
+            inner_dict["prev"] = "night-"+f[i-1]+".html"
+            inner_dict["next"] = "night-"+f[i+1]+".html"
+        file_js["n"+f[i]] = inner_dict
+
+    outfile = os.path.join(outdir, 'linking.js')
+    with open(outfile, 'w') as fp:
+        fp.write("get_dict({})".format(str(file_js)))
```

Comment:
This seems a little magical to me.  Is it relying upon the fact that python dictionaries represented as strings are extremely similar to json?  I think they differ by single vs. double quote usage and browsers might be forgiving of that, but better to use `json.dumps(file_js)`.  Or maybe I'm completely missing how this works.